src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.State.cs

```diff
@@ -314,8 +315,12 @@ private State()
 
                 bool itemsChanged = false;
 
-                foreach ((string itemType, IProjectChangeDescription projectChange) in sourceItemsUpdate.ProjectChanges)
+                foreach ((string schemaName, IProjectChangeDescription projectChange) in sourceItemsUpdate.ProjectChanges)
                 {
+                    // ProjectChanges is keyed by the rule name which is usually the same as the item type, but not always (eg, in auto-generated rules)
+                    string? itemType = projectCatalogSnapshot.NamedCatalogs[PropertyPageContexts.File].GetSchema(schemaName)?.DataSource.ItemType;
```

Comment:
Is this indexer guaranteed to find its key?

Reply:
There is certainly lots of spots in code where it's assumed to exist, but an extra question mark here wouldn't hurt much. 👍